### ISSUE TITLE
[feature] string quote escaping for cmake defs

### ISSUE THREAD

**Issue by EmingK** [conan-io/conan#7992]
Some CMake definitions contains characters that need to be quoted to be correctly explained. For example,

``` sh
# the command line below breaks.
# cmake explains it as key="CMAKE_XCODE_ATTRIBUTE_EXCLUDED_ARCHS[", value="iphoneos*]=armv7s"
cmake -DCMAKE_XCODE_ATTRIBUTE_EXCLUDED_ARCHS[sdk=iphoneos*]=armv7s

# the command line below works.
cmake '-D"CMAKE_XCODE_ATTRIBUTE_EXCLUDED_ARCHS[sdk=iphoneos*]"=armv7s'
```

In `conanfile.py` DSL, `defs` arg in `CMake.configure()` pass defs dictionary as command line args to cmake. If definitions like the example above is provided, you must write it in a weird way to make things work:

``` python
    def build(self):
        generator = "Unix Makefiles"
        cmake_defs = {}
        if self.settings.os == "iOS":
            generator = "Xcode"
            cmake_defs["CMAKE_SYSTEM_NAME"] = "iOS"
            cmake_defs["CMAKE_IOS_INSTALL_COMBINED"] = "TRUE"
            cmake_defs["CMAKE_OSX_ARCHITECTURES"] = "armv7 arm64 x86_64"
            cmake_defs["CMAKE_OSX_DEPLOYMENT_TARGET"] = self.settings.os.version

            # you must write definitions containing `=` character like this to make things work!
            # the outer quotation is for shell, the inner is for cmake
            cmake_defs["'\"CMAKE_XCODE_ATTRIBUTE_EXCLUDED_ARCHS[sdk=iphonesimulator*]\"'"] = "arm64"

        cmake = CMake(self, generator=generator)
        cmake.configure(defs=cmake_defs)
        cmake.build(target="install")
```

It can be better if keys in `defs` can be automatically escaped, and we can simply write 

```
cmake_defs["CMAKE_XCODE_ATTRIBUTE_EXCLUDED_ARCHS[sdk=iphonesimulator*]"] = "arm64"
```

without considering character escaping.

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by Johnnyxy** [comment]
Hi there,
as side note: I have seen the usage of square brackets in CMake. But it is discouraged:
https://cmake.org/cmake/help/latest/manual/cmake-language.7.html#variables

As for your feature request, I think it would generally be advised to quote commandline arguments.
This eleminates any ambiguity of commandline parameters. Parsing of the commandline parameters does not need special handling of quotes in values.
```
cmake ... "-DABC=def"
cmake ... "-DABC=def ghi"
cmake ... "-DABC=\"def ghi klm\""
```

As reference for the command line on Windows:
On Windows the quotes in environment variables could be interpreted as content of the value (see [here](https://ss64.com/nt/set.html)).
```
set ABC="def ghi"
echo %ABC%
set "ABC=def ghi"
echo %ABC%
```
results:
```
"def ghi"
def ghi
```

**Comment by EmingK** [comment]
@Johnnyxy brackets not matter but the equal character (`=`).

The example mentioned in the issue is pretty common in Apple platform (i.e. building on OSX or cross-compiling to iOS), of which the definition name contains a `=` character:

``` shell
cmake ... -DA=B=C
# is equivalent to
cmake ... '-DA="B=C"'
# but differs from
cmake ... '-D"A=B"=C'
```

with current version,
``` python
# what you write
cmake_defs["A=B"] = "C"
# what you actually get
cmake_defs["A"] = "B=C"
```

**Comment by Johnnyxy** [comment]
Oh I forgot to add the following :).

Your Scenario
--------------
When you do the following
```cmake_defs["A[B]"] = "C"``` results in the command line portion ```-DA[B]="C"``` and the ```main``` function argv[...] = ```-DA[B]="C"```. In theory CMake parses this as variable ```A[B]``` with the content ```C``` (CMake's parsing strips away the quotes).

I tried the following:
```python
cmake = CMake(self)
cmake.definitions["'\"CMAKE_XCODE_ATTRIBUTE_EXCLUDED_ARCHS[sdk=iphonesimulator*]\"'"] = "arm64"
print (cmake.command_line)
```
results in command line portion
```
-D'"CMAKE_XCODE_ATTRIBUTE_EXCLUDED_ARCHS[sdk=iphonesimulator*]"'="arm64"
```
CMake was unable to parse this as it sees only ```'CMAKE_XCODE_ATTRIBUTE_EXCLUDED_ARCHS[sdk``` as variable name.
When I tried to print the content of this name CMake errors out:
```
message (STATUS "${'CMAKE_XCODE_ATTRIBUTE_EXCLUDED_ARCHS[sdk}")
```
```
CMake Error at CMakeLists.txt:25 (message):
  Syntax error in cmake code at

    D:/test/CMakeLists.txt:25

  when parsing string

    ${'CMAKE_XCODE_ATTRIBUTE_EXCLUDED_ARCHS[sdk}

  Invalid character (''') in a variable name: ''

```
If you leave out the single quotes then you will get the same error but for the square bracket.
This results to the conclusion: ```You cannot use square brackets or other special characters in variable names.```

Conan - Quoting
------------------
What I was trying to say is that one should always quote command line arguments entirely.
Because what the command line parser (bash, zsh, Windows cmd.exe, ...) parses and what the executed process parses can differ. As the process' parsing is entirely up to the process' parsing rules.

```
1. cmake ... -DABC="def"
2. cmake ... "-DABC=def"
```
results in the following argument element in the ```main``` function's ```argv``` parameter
```
1. argv[...] = "-DABC=\"def\""
2. argv[...] = "-DABC=def"
```
What CMake is doing now is parsing it with its own logic to parse the arguments and throws away the quotes in the first example.
So CMake is able to see the values as equal but this is up to CMake's parsing rules. Thus quoting the entire argument should be preferred. This works with Windows commandline and at least Unix'/Linux' bash.

CMake variable names
-----------------------
The variable names in CMake should only consist of text characters, numbers, underscores and dashes. Square brackets will lead to errors in CMake.

Quote from the [CMake documentation](https://cmake.org/cmake/help/latest/manual/cmake-language.7.html#variables):
```Variable names are case-sensitive and may consist of almost any text, but we recommend sticking to names consisting only of alphanumeric characters plus _ and -.```

**Comment by EmingK** [comment]
@Johnnyxy thanks for replying!

## About brackets in variable name

CMake does support it in variable name (though not encouraged), but just cannot use it "literally". With double substitution of variable name, you get access to it.

You can try this:

``` cmake
project(test)

# error!
# message(
#     STATUS
#     "${A_VAR_WITH_[BRACKETS]}"
# )

# ok!
set(VAR_NAME "A_VAR_WITH_[BRACKETS]")
message(
    STATUS
    "${${VAR_NAME}}"
)
```
with
``` sh
cmake . "-DA_VAR_WITH_[BRACKETS]=Hello"
```

## About my scenario

It is fundamentally irrelavent to brackets. It's about the equal `=` character.

Just throw the brackets away and I'll not use them anymore after this line :) 

Now I would like to define a cmake variable, of which the variable name is `A=B` and value is `C`. 

Note that variable is also not directly expressible in CMake, printing it requires the same way the CMake example above does ._.

I'll write

``` python
cmake_defs["A=B"] = "C"
```

Conan simply join the key with `=`, producing command line
``` sh
-DA=B=C
```
which is interpreted by cmake as variable name is `A`, value is `B=C`.

To actually define the variable I want, the command line should be 
``` sh
-D"A=B"=C
```

## About quoting in command line / shells

I'm using PowerShell under macOS. Actual command line may vary depending on platform / shell. Therefore, escaping the same commandline can be different under different platforms.

For bash/zsh/... it should use
``` shell
cmake ... "-D\"A=B\"=C"
# or
cmake ... '-D"A=B"=C'
```

For cmd
``` batch
cmake ... -D\"A=B\"=C
```

Platform differences make writing `conanfile` with correct manual escaping more difficult. Generating correct escaping within conan framework eases package developers' mental burden :)

**Comment by Johnnyxy** [comment]
About brackets in variable name
---------------------------------
WTF :). Yes this is a workaround. When I played around with variable names I did not test such an approach. Nice to know this "hack".

Your Scenario
--------------
Yes, in your scenario the problem is the equal sign. With the variable name workaround you mentioned this would lead to your described problem.

Quoting
---------
It was more of a general advise and propably a little bit too unspecific :).

On Windows one should quote each single argument entirely. This ensures that the executed process does not have to deal with special characters like spaces etc.
```programm.exe "arg1" "-arg2=2" "-arg3 3" "-arg4 \"4\""```
```
arg1
-arg2=2
-arg3 3
-arg4 "4"
```

But on *nix systems one can use escaping to ensure one argument does not get split into multiple ```argv's```.
```programm.exe arg1 -arg2=2 -arg3\ 3 -arg4\ \"4\"```

CMake - Quoting
-----------------
In this specific case we are talking about executing a CMake process what does have special quotation handling. So Conan could do it like this on Windows:
```
cmake.exe ... "-D\"NAME\"=a b" ...
```
On other systems:
```
cmake ... -D"NAME"=a\ b ...
```
